Code Challenge -  (Carlos Henrique)

RickZzDev opened the pull request:
Thanks for the opportunity guys, this test was super cool.

I hope you guys like it.

Obs: The GraphQL request is mocked due to the daily limit.  But the real implementation is there two,, if you want use just change de dependency injection on restaurant_module

spencer-sf commented on test/features/restauran_tour_test/restauran_tour_test.dart:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  testWidgets('All restaurant list loaded', (WidgetTester tester) async {
+    /// Load the app with material app configuration
```
**question**: Why doc comments?

spencer-sf commented on test/features/restauran_tour_test/restauran_tour_test.dart:
```diff
@@ -0,0 +1,72 @@
+    /// check at end of list, should display 'View More' text button
+    expect(find.text("View More"), findsOneWidget);
+
+    /// Identify 'My Favorites' tab and tap on it
+    Finder myFavoriteTab = find.widgetWithText(Tab, "My Favorites");
+    await tester.tap(myFavoriteTab);
+    await tester.pumpAndSettle();
```
**suggestion**: Tests should be as small as possible, testing only one component in each test. Consider breaking these into separate tests

spencer-sf commented on test/features/restauran_tour_test/restauran_tour_test.dart:
```diff
@@ -0,0 +1,72 @@
+
+    /// Check if initial screen is laoded with two tab bars as 'All Resturants' and 'My Favorites'
+    expect(find.byType(TabBar), findsOneWidget);
+    expect(find.widgetWithText(Tab, "All Resturants"), findsOneWidget);
```
**nit-pick**: Changing the text on the tab in the UI will cause this test to break, move to a constant pattern to avoid test brittleness

spencer-sf commented on test/features/restauran_tour_test/restauran_tour_test.dart:
```diff
@@ -0,0 +1,72 @@
+  });
+
+  ///TODO: Navigation tetsing is failing need to fix it.
+  // testWidgets('open restauran detail screen', (WidgetTester tester) async {
```
**nit-pick**: Tests are equally as important as features, don't merge code with failing/commented out tests

spencer-sf commented on test/features/restauran_tour_test/restauran_tour_test.dart:
```diff
@@ -0,0 +1,72 @@
+
+  ///TODO: Navigation tetsing is failing need to fix it.
+  // testWidgets('open restauran detail screen', (WidgetTester tester) async {
+  //   await _createWidget(tester);
```
**bug-fix**: When inflating the material widget to test, you have to have a route handler to test navigation so mocking one out will fix this test

spencer-sf commented on test/features/restauran_tour_test/restaurants_view_model_test.dart:
```diff
@@ -0,0 +1,58 @@
+    restauranService = MockRestauranService();
+  });
+
+  testWidgets('Check all restautant list got loaded',
```
**question**: Since all of these functions provide only business logic, can they be tested with a unit test instead of a widget test?

enzo-sf commented on lib/app_routes/app_routes.dart:
```diff
@@ -0,0 +1,16 @@
+
+  static setUpRoutes(BuildContext context) {
+    return <String, WidgetBuilder>{
+      initialRoute: (BuildContext context) => const RestauranTour(),
+      restaurantDetailRoute: (BuildContext context) =>
+          const ResturantDetailScreen(),
```
suggestion: In both cases, context is not used. So it could be removed to improve readability.

```suggestion
      initialRoute: (_) => const RestauranTour(),
      restaurantDetailRoute: (_) => const ResturantDetailScreen(),
```

enzo-sf commented on lib/features/restauran_tour/model/restauran_service.dart:
```diff
@@ -0,0 +1,78 @@
+
+  Future<Object> getAllRestaurans({int offSet = 0}) async {
+    try {
+      var response = await yelpRepo.getResponse(
+        path: restauranUrl,
+        query: _getQuery(offSet),
+      );
```
suggestion: prefer `final` over `var` if the value is not going to be updated.

enzo-sf commented on lib/repositories/endpoints.dart:
```diff
@@ -0,0 +1,4 @@
+const String yelpKey =
+    "98_32Ccpik5XnkA-c3yz6ifRB7tJTeyuqXWif8Lpwt3SYUnPEfOlDwJDRMW3Gv9_XJNbSPPD1hsm0ugp81Zk6-gbxsjFXN5OC2Hyta5o_bA2IjqbN_exI0AYHW2rYnYx";
+const String baseUrl = "https://api.yelp.com";
+const String restauranUrl = "/v3/graphql";
```
typo: restaurant

enzo-sf replied:
typo: Also in `restaurant_service.dart`

enzo-sf commented on lib/features/restauran_tour/model/restauran_service.dart:
```diff
@@ -0,0 +1,78 @@
+        path: restauranUrl,
+        query: _getQuery(offSet),
+      );
+      if (200 == response.statusCode) {
```
I know it's the same but it's a personal preference
```suggestion
      if (HttpStatus.ok == response.statusCode) {
```

enzo-sf commented on lib/features/restauran_tour/model/restauran_service.dart:
```diff
@@ -0,0 +1,78 @@
+    }
+  }
+
+  String _getQuery(int offset) {
```
suggestion: improve naming (like `_getRestaurantsQuery`)

enzo-sf commented on lib/features/restauran_tour/model/restauran_service.dart:
```diff
@@ -0,0 +1,78 @@
+query getRestaurants {
+  search(location: "Las Vegas", limit: 20, offset: $offset) {
+    total    
+    business {
+      id
+      name
+      price
+      rating
+      photos
+      reviews {
+        text
+        id
+        rating
+        user {
+          id
+          image_url
+          name
+        }
+      }
+      categories {
+        title
+        alias
+      }
+      hours {
+        is_open_now
+      }
+      location {
+        address1
+        address2
+        address3
+        city
+        state
+        postal_code
+        country
+        formatted_address
+      }
+    }
+  }
+}
+''';
```
thought: One of the key arguments in favor of graphql is that we don't need to fetch information that will not be required for that specific view. If you take a look into the restaurant item view you don't need `reviewers` 
and `location`.

<img width="479" alt="image" src="https://user-images.githubusercontent.com/95098927/177221856-b1564521-02cd-4db7-aed6-6fa8f0524ae8.png">

enzo-sf replied:
Also some of the requirements it's the loading state for the detail screen. So probably my suggestion it's to split this into 2 different queries one for the restaurant list and the other one for detail.

<img width="408" alt="image" src="https://user-images.githubusercontent.com/95098927/177222113-624ef5a8-3fe5-4f29-a09c-aa8cffaa8128.png">

enzo-sf commented on lib/features/restauran_tour/view/restaurant_detail_screen.dart:
```diff
@@ -0,0 +1,204 @@
+          ),
+        ],
+      ),
+      body: SingleChildScrollView(
+        child: Column(
+          children: [
+            _imageDisplay(context, viewModel.selectedRestaurant.heroImage),
+            _detailContent(context, viewModel.selectedRestaurant),
+          ],
+        ),
+      ),
```
suggestion: Instead of using 2 nested widgets take advantage of the more accurated

```suggestion
    body: ListView(
        children: [
          _imageDisplay(context, viewModel.selectedRestaurant.heroImage),
          _detailContent(context, viewModel.selectedRestaurant),
        ],
      ),
```

enzo-sf commented on lib/features/restauran_tour/view_model/restaurants_view_model.dart:
```diff
@@ -0,0 +1,94 @@
+import 'package:restaurantour/repositories/api_status.dart';
+
+class RestaurantsViewModel extends ChangeNotifier {
+  RestauranService? service;
```
suggestion: Instead of declaring this as nullable (this would never happen) make it `late` to initialize it later

```suggestion
  late RestauranService service;
```

enzo-sf commented on lib/features/restauran_tour/view_model/restaurants_view_model.dart:
```diff
@@ -0,0 +1,94 @@
+    getAllRestaurants(offSet: _offSet, isRefresh: false);
+  }
+
+  void setSelectedRestaurant(Restaurant item) {
+    _selectedRestaurant = item;
+  }
```
Instead of handling this on memory, you should make here the query to fetch the detail as I mentioned before

enzo-sf commented on lib/widgets/rating.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:restaurantour/theme/app_color.dart';
+
+class Rating extends StatelessWidget {
+  final double? rating;
```
thought: I think that this does not need to be nullable.

enzo-sf commented:
improvement-bug: Set a fixed height for the image on the detail screen.

<img width="476" alt="image" src="https://user-images.githubusercontent.com/95098927/177223418-0c2d4441-eb5f-4ac1-807f-0a3b0094091c.png">

enzo-sf commented:
Hi @prashanthvontela, very nice challenge! As @spencer-sf mentioned we take a review as it would be a real client project, so every suggestion it's made with love ✌️.

Also, I want to second Spencer in some suggestions so you can take care of it in the future.
- Use Widget (stless) instead of functions that return widgets
- Don't return `Object`
- Avoid passing whole objects just for single fields
- Keep your test as small as possible

Commit:
chore: initial project setup

- create initial project structure
- create package of dependency injection
- create package of ui ki
- add melos to deal with mono repo structure

Commit:
feat: Add restaurants list page and details

- Add page that list all restaurants
- Add page dedicated to show restaurant deails

Commit:
feat: Add possibility to add restaurant as a favorite

- create favorite restaurant list tab
- add funcionality to hear icon that adds restaurant as favorite

Commit:
test: Add tests of feature and platform most relevant packages

Commit:
feat: Add error body

- Create widget to be shown when view receives and error from presenter

Commit:
fix: Add TextOverflow no restaurant name

- Fix the restaurant name on restaurant card that was overflowing when it was too wide

Commit:
fix: Fixing favorite add and remove

- fix favorite behavior, when user unfavorite the restaurant the tab reload when it is popped
- fix restaurant isFavorite property, now it is on toJson em fromJson

Commit:
test: Add end to end testing

- Add end to end flows of maestro
- Add video of testing

Commit:
doc: Add project documentation

- Add README of the project
- Add screenshots of documentatios

RickZzDev committed:
doc: Add screenshots to ReadME

Commit:
doc: Improve doc

- add better screenshot title

Commit:
fix: Fix addition of more than one favorite

- Fix bug when user added more than one restaurant as favorite, it was adding only one
- Test the bug fix with end to end and unit tests

RickZzDev committed:
doc: Add final End to End video to README

Commit:
doc: Improve docs

- Add command to run, and how to install melos

Commit:
chore: Cleaning launch.json

Commit:
fix: Remove hard coded image

Commit:
chore: Testing and refactor restaurant repository

- Create mocked repository due to daily limit
- Rename yelp repository
- Add tests
